### PR TITLE
feat(update/registry): remote update from ECR private registry with cert [EE-5164] 

### PR DESCRIPTION
**Pull request by oscarzhou-portainer**
closes [EE-5164]

[EE-5164]: https://portainer.atlassian.net/browse/EE-5164?

### CODE REVIEW

File: edge/yaml/docker_compose.go [hunk 1]
```diff
@@ -57,30 +51,77 @@ func addEnvsForSpecificService(fileContent, serviceName string, envs map[string]
...
+		if err != nil {
+			return "", err
+		}
+
+		envs["REGISTRY_USED"] = "1"
+		// hardcode username for aws ecr registry
+		// @https://docs.aws.amazon.com/cli/latest/reference/ecr/get-login-password.html#examples
+		envs["REGISTRY_USERNAME"] = "AWS"
+		envs["REGISTRY_PASSWORD"] = c.Secret
+		envs["REGISTRY_PASSWORD"] = c.Secret
```
Comment (chiptus) [inline edge/yaml/docker_compose.go]: remove duplicate line

File: edge/yaml/docker_compose_test.go [hunk 2]
```diff
@@ -1,109 +1,84 @@
...
-			name: "valid docker compose file content",
-			input: `version: "3"
-services:
-invalid_service:`,
-			expected: ``,
-			wantErr:  true,
+			name:      "empty image name",
+			imageName: "",
+			expected:  "",
+			err:       errors.New("No image name provided"),
```
Comment (chiptus) [inline edge/yaml/docker_compose_test.go:99]: is something like `http://localhost:3000/image:tag` a possible registry? and what about `localhost:3000/image:tag`?

File: edge/yaml/docker_compose.go [hunk 3]
```diff
@@ -57,30 +51,77 @@ func addEnvsForSpecificService(fileContent, serviceName string, envs map[string]
...
+
+				envs["REGISTRY_USED"] = "1"
+				envs["REGISTRY_USERNAME"] = cred.Username
+				envs["REGISTRY_PASSWORD"] = cred.Secret
+				break
+			}
+		}
+	}
+
+	return updateServiceWithEnv(compose, serviceName, envs)
```
Comment (chiptus) [inline edge/yaml/docker_compose.go:97]: maybe we should think builder pattern. so we call this:
```
ym := yaml.NewDockerComposeYaml(....); // here we already have parsed ym as `ym.parsed` or `ym.data`
ym.WithCredentails("updater")
ym.WithEnv("updater", envs)
newFile := ym.String()
```

wdyt?

File: edge/aws/aws.go [hunk 4]
```diff
@@ -71,3 +76,18 @@ func authenticateAgainstIAMRA(awsConfig *agent.AWSConfig) (*iamra.CredentialProc
 
 	return &credentialProcessOutput, nil
 }
+
+func ExtractAwsConfig(options *agent.Options) *agent.AWSConfig {
+	if os.IsValidAWSConfig(options) {
```
Comment (chiptus) [inline edge/aws/aws.go]: maybe move IsValidAWSConfig here?

File: edge/stack/stack.go [hunk 5]
```diff
@@ -512,8 +514,8 @@ func (manager *StackManager) buildDeployerParams(stackData client.EdgeStackData,
 
 	switch manager.engineType {
 	case EngineTypeDockerStandalone, EngineTypeDockerSwarm:
-		if len(stackData.RegistryCredentials) > 0 && strings.HasPrefix(stackData.Name, "edge-update-schedule") {
-			yml := yaml.NewDockerComposeYAML(fileContent, stackData.RegistryCredentials)
+		if (len(stackData.RegistryCredentials) > 0 || manager.awsConfig != nil) && strings.HasPrefix(stackData.Name, "edge-update-schedule") {
```
Comment (chiptus) [inline edge/stack/stack.go]: tech debt: change the last condition from `strings.HasPrefix(stackData.Name, "edge-update-schedule")` to `stackData.AddCredentials`

File: edge/yaml/docker_compose_test.go (hunk 2 above)
Comment (chiptus) [reply edge/yaml/docker_compose_test.go:99]: just to be clear, I'm not sure if those are valid registries, but if they are, we need to take care of them

**Review by chiptus — COMMENTED** [review]
mostly suggestions, need to remove duplicate line and see that we take care of all registry types.

File: edge/yaml/docker_compose_test.go (hunk 2 above)
Comment (oscarzhou-portainer) [reply edge/yaml/docker_compose_test.go:99]: I'm surprised to see how many valid image URL formats there are. To ensure we have adequate test coverage, I will be adding more unit test cases.

File: edge/yaml/docker_compose.go (hunk 3 above)
Comment (oscarzhou-portainer) [reply edge/yaml/docker_compose.go:97]: If we're only working with Docker Compose YAML files, I don't think we need to create a pattern right away. Alternatively, we could consider developing a pattern that covers both Docker Compose YAML and K8s YAML in a separate project to make the workflow more clear and organized.

File: edge/stack/stack.go (hunk 5 above)
Comment (oscarzhou-portainer) [reply edge/stack/stack.go]: Can you please explain this more?
This registry credential injection here is only for remote update from private registry

File: edge/yaml/docker_compose_test.go (hunk 2 above)
Comment (chiptus) [reply edge/yaml/docker_compose_test.go:99]: maybe we can use https://github.com/distribution/distribution/blob/v2.7.1/reference/reference.go#L189 ?

File: edge/stack/stack.go (hunk 5 above)
Comment (chiptus) [reply edge/stack/stack.go]: it's very specific. what if we change the name of the stack? this should be an edge stack property and not some magic string check

**Review by chiptus — APPROVED** [review]

File: edge/registry/server.go [hunk 6]
```diff
@@ -37,7 +38,7 @@ func NewEdgeRegistryHandler(edgeManager *edge.Manager, awsConfig *agent.AWSConfi
 func (handler *Handler) LookupHandler(rw http.ResponseWriter, r *http.Request) *httperror.HandlerError {
 	stackManager := handler.EdgeManager.GetStackManager()
 	if stackManager == nil {
-		return &httperror.HandlerError{http.StatusInternalServerError, "Unable to retrieve stack manager", errors.New("Stack manager is not available")}
+		return httperror.InternalServerError("Unable to retrieve stack manager", errors.New("Stack manager is not available"))
```
Comment (chiptus) [inline edge/registry/server.go:41]: please apply these changes to develop version

**Review by chiptus — CHANGES_REQUESTED** [review]
apply changes to develop

File: edge/registry/server.go (hunk 6 above)
Comment (oscarzhou-portainer) [reply edge/registry/server.go:41]: Someone already made the changes to the develop branch before I could submit my pull request.

**Review by chiptus — APPROVED** [review]
LGTM

**Pull request merged**